src/praisonai-agents/praisonaiagents/llm/llm.py

```
@@ -432,13 +437,163 @@ def _supports_xml_tool_format(self) -> bool:
+                if attempt < self._max_retries:
+                    retry_delay = self._parse_retry_delay(error_str)
+
+                    logging.warning(
+                        f"Rate limit hit (attempt {attempt + 1}/{self._max_retries + 1}), "
+                        f"waiting {retry_delay:.1f}s before retry..."
+                    )
+
+                    if self._rate_limiter is not None:
+                        await self._rate_limiter.wait_for_retry_async(retry_delay)
+                    else:
+                        await asyncio.sleep(retry_delay)
```

Ruff (0.14.10)

Undefined name `asyncio`

(F821)